### PR TITLE
Remember record

### PR DESCRIPTION
**Pull request by BlckHawker**
[Ticket](https://github.com/orgs/RIT-Crowd-Comic/projects/2/views/1?pane=issue&itemId=63814608)

### Description
- Use `/start-remembering` to start remembering messages in a specific channel

#### Parameters
- `channel-id` (optional) - The id of the channel the messages are in. If not provided, default is the channel the command was ran
- `exclude-bot-messages` (optional) - If bot messages should be excluded in the message collection. If not provided, default is used from `config.json` (`defaultExcludeBotMessages.startRemember`)

### Notes
- If `/start-remembering` is called twice without `/stop-remembering` being called in between, an error message will be sent, stating to call `/stop-remembering`to stop remembering messages in the specified channel
- When `/start-remembering`, the bot's status will update to say which channel's messages are being remembered

### Screenshot(s)
image

### Known Issues / Warnings
- If the bot goes offline after the `/start-remembering`is ran, the bot will act like the command was not ran when it goes up again. If wanted, it's possible to add this via the DB once it's implemented. Could be useful if the bot goes down for maintenance.

### Possible Additional Features
- Currently there is no way to have multiple remembered calls remembered. It could be possible with the usage ids. This could be beneficial if people want to remember conversations in different channels. However, our current schema of remembering messages has no way to separate channel conversations. Causing all messages to be combined in to one conversation.
- Currently, there is no way to change the ephemeral via bot commands; only possible via `ephemeral.startRemember` in `config.json`. If requested, this can be changed, but don't see the practicality of it.

### Description
- Use `/stop-remembering` to stop remembering messages from a specified channel. All the messages that were between the `/start-remembering` and `/stop-remembering` commands will saved.

#### Parameters
N/A

### Screenshot(s)
image

### Notes

- If the command is ran before `/start-remembering`, a warning message will appear, stating to run `/start-remembering` first

### Known Issues / Warnings
N/A

### Possible Additional Features
N/A

### CODE REVIEW

File: src/commands/remember/recall.js [hunk 1]
```diff
@@ -1,24 +1,20 @@
...
         try {
             //get the json
             //? There is probably a way to do this so we don't have to have a fle created to save to
             //? Will probably be resolved once DB is created
             await interaction.deferReply({ ephemeral: false })
             const jsonFilePath = './src/rememberedMessages.json';
-            const messageObj = []
-            const messages = getRememberedMessage();
-            messages.forEach(m => messageObj.push(m))
+            const messages = structuredClone(getRememberedMessage());
```
Comment (tstodter) [inline src/commands/remember/recall.js:15]: This is really cool; I didn't know about `structuredClone`, so 👍 .

Also, I still think you don't need to clone `getRememberedMessage()` at all; you only use `messages` in the `stringify` call, which does not mutate it. So can just use `getRememberedMessage()` directly. If this is here out of an abundance of caution, then I think that's well-placed...but I might suggest just doing the clone inside of `getRememberedMessage`, so no client of that function can modify the data without going through `addMessage`.

File: src/commands/remember/rememberRangeGrab.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 const { ApplicationCommandOptionType, PermissionFlagsBits } = require('discord.js');
-const remeberMessagesMethods = require("../../utils/rememberMessages");
+const remeberMessagesUtils = require("../../utils/rememberMessages");
```
Comment (tstodter) [inline src/commands/remember/rememberRangeGrab.js:2]: We should fix this spelling; I'm counting 14 instances of "remeber" in this PR.

**Review by tstodter — APPROVED** [review]
LGTM, untested.

**Comment by tstodter** [comment]
What are the team's thoughts on a small refactor of the /remember* commands? To switch from something like `/start-remembering` and `/stop-remembering` and `/remember-chunk-time` etc, and moving to `/remember start` and `/remember stop` and `/remember chunk-time`? Could use [subcommands](https://discordjs.guide/slash-commands/advanced-creation.html#subcommands) for this.

File: src/commands/remember/rememberRangeGrab.js (hunk 2 above)
Comment (BlckHawker) [reply src/commands/remember/rememberRangeGrab.js:2]: These have been fixed in [this PR](https://github.com/RIT-Crowd-Comic/Bot/pull/14). Hoping doing it like this decreases the chances of merge conflicts

File: src/commands/remember/recall.js (hunk 1 above)
Comment (BlckHawker) [reply src/commands/remember/recall.js]: Updated so `getRememberedMessage` returns a clone

**Pull request merged**